### PR TITLE
chore(sdk)!: create oracle client interface for new oo contract integrations

### PR DESCRIPTION
**Pull request by daywiss**
BREAKING CHANGE: this slightly changes the request type which is used by the UI, app will need to be
adjusted to accomodate.

Signed-off-by: David <david@umaproject.org>

**Motivation**
We want to be able to support multiple oracle contracts. 

**Summary**
Originally i made  pr that altered the oracle client in a different way to support new oracle contracts. This is a different approach which isolates changes to the oracle service itself. This makes an oracle interface which needs to be implemented. This alters the design of the client somewhat, now the oracle clients are stateful now rather than stateless. Effort is taken not to break existing logic, but some types are slightly altered to simplify typing.  The existing functionality has not changed, only organization of the code. 

**Testing**

 This was tested by integrating it into the UI and looking for regressions on kovan. 

- [x]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/sdk/src/clients/skinnyOptimisticOracle/client.ts [hunk 1]
```diff
@@ -93,12 +93,12 @@ export function parseRequest(request: SolidityRequest) {
...
-    reward: request.proposedPrice.toString(),
-    finalFee: request.proposedPrice.toString(),
-    bond: request.proposedPrice.toString(),
-    customLiveness: request.proposedPrice.toString(),
+    resolvedPrice: request.resolvedPrice.toString(),
+    expirationTime: Number(request.expirationTime.toString()),
+    reward: request.reward.toString(),
+    finalFee: request.finalFee.toString(),
+    bond: request.bond.toString(),
+    customLiveness: request.customLiveness.toString(),
```
Comment (daywiss) [inline packages/sdk/src/clients/skinnyOptimisticOracle/client.ts:101]: this was a bug, fortunately not used yet

File: packages/sdk/src/oracle/client.ts [hunk 2]
```diff
@@ -162,6 +164,19 @@ export class Client {
   }
 }
 
+function makeProvider(rpcUrls: string[]): FallbackProvider {
```
Comment (daywiss) [inline packages/sdk/src/oracle/client.ts:167]: this logic was pulled out of the write class unchanged for the most part. just so I can use the provider during configuration

File: packages/sdk/src/oracle/services/optimisticOracle.ts [hunk 3]
```diff
@@ -1,66 +1,108 @@
...
+import { RequestPrice, ProposePrice, DisputePrice, Settle } from "../../clients/optimisticOracle";
 
-export class OptimisticOracle {
-  public readonly contract: optimisticOracle.Instance;
-  constructor(protected provider: Provider, protected address: string) {
+export type OptimisticOracleEvent = RequestPrice | ProposePrice | DisputePrice | Settle;
+
+export type { RequestPrice, ProposePrice, DisputePrice, Settle };
+
+export class OptimisticOracle implements OracleInterface {
```
Comment (daywiss) [inline packages/sdk/src/oracle/services/optimisticOracle.ts:12]: this class is the bulk of the new logic changes. it updates the oracle client to be stateful, and handle a lot of its own processing internally. It conforms to the oracle interface.

File: packages/sdk/src/oracle/types/interfaces.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { BigNumberish, BigNumber, Signer, TransactionResponse, TransactionReceipt } from "../types/ethers";
```
Comment (daywiss) [inline packages/sdk/src/oracle/types/interfaces.ts:1]: this file defines the interfaces that are needed to generalize the oracle service

**Review by chrismaree — COMMENTED** [review]
I dont really see unit tests here? perhaps this is part of the SDK that is not tested? did you try this out in the production OO UI to verify the correctness in prod?

**Comment by daywiss** [comment]
@chrismaree I will double check that the end to end tests work, but i primarily tested this by integrating it into the frontend. I wouldnt suggest doing the integration after this upgrade though and wait until the next pr where we will add the skinny oo.

**Comment by daywiss** [comment]
@chrismaree updated end to end tests that were failing.

**Review by mrice32 — APPROVED** [review]
LGTM!

**Pull request merged**